src/metabase/query_processor/middleware/cache_backend/db.clj

```diff
@@ -48,25 +51,49 @@
       (doto stmt
         (.setFetchDirection ResultSet/FETCH_FORWARD)
         (.setBytes 1 query-hash)
-        (.setObject 2 (seconds-ago max-age-seconds) Types/TIMESTAMP_WITH_TIMEZONE)
+        (.setObject 2 updated-at Types/TIMESTAMP_WITH_TIMEZONE)
         (.setMaxRows 1))
       (catch Throwable e
-        (log/error e (trs "Error preparing statement to fetch cached query results"))
+        (log/error e "Error preparing statement to fetch cached query results")
         (.close stmt)
         (throw e)))))
 
-(defn- cached-results [query-hash max-age-seconds respond]
+
+(defn fetch-cache-stmt-ttl
+  "Make a prepared statement for :ttl caching strategy"
+  ^PreparedStatement [strategy query-hash ^Connection conn]
+  (when (and (= :ttl (:type strategy))
+             (:execution-time strategy)) ;; no cache when it was never executed before
+    (let [max-age-seconds (math/round (/ (* (:multiplier strategy)
+                                            (:execution-time strategy))
+                                         1000.0))]
+      (prepare-statement conn query-hash (seconds-ago max-age-seconds)))))
+
+(defenterprise fetch-cache-stmt-ee
+  "Returns prepared statement for a given strategy and query hash - on EE. Returns `::oss` on OSS."
+  metabase-enterprise.caching.strategies
+  [_strategy _conn _hash]
+  ::oss)
+
+(defn- fetch-cache-stmt ^PreparedStatement [strategy hash conn]
+  (let [res (fetch-cache-stmt-ee strategy hash conn)]
+    (if (= res ::oss)
```

Comment:
this feels unnecessary to me, I think you can have `fetch-cache-stmt` as a `defenterprise` and make :oss version implement `fetch-cache-stmt-ttl`

Reply:
indeed, makes for a bit simpler code (but `defenterprise` doesn't keep type hints ATM)